## Recursão

Problema: calcular fatoração "n! = n * (n-1)!".

In [1]:
def fat(n):
    if n == 0:
        return 1
    else:
        res = n * fat(n-1)
        return res

In [2]:
fat(4)

24

Problema: n-ésimo termo de Fibonacci.

In [3]:
def fib_rec(n):
    if n <2:
        return n
    else:
        return fib_rec(n-1) + fib_rec(n-2)

In [4]:
fib_rec(4)

3

## Fibonacci Recursivo x Finonacci Iterativo

In [13]:
import time 

In [16]:
def fib_rec(n):
    if n <2:
        return n
    else:
        return fib_rec(n-1) + fib_rec(n-2)
    

def fib_it(n):
    res = n
    a, b = 0, 1
    for k in range(2, n+1):
        res = a + b
        a, b = b, res
    return res


n = 30
start = time.time()
print(fib_rec(n))
print('Recursive: {} seconds'.format(time.time() - start))

start = time.time()
print(fib_it(n))
print('Iterative: {} seconds'.format(time.time() - start))


832040
Recursive: 1.7807822227478027 seconds
832040
Iterative: 0.0 seconds


## Memoização

Memoização do n-ésimo termo da sequência de Fibonacc.

In [18]:
m = dict()
def fib_mem(n):
    if n < 2:
        return n
    elif m.get(n) != None:
        return m[n]
    else:
        m[n] = fib_mem(n-1) + fib_mem(n-2)
        return m[n]
    

n = 30
start = time.time()
print(fib_mem(n))
print('Memoization: {} seconds'.format(time.time() - start))

832040
Memoization: 0.0009989738464355469 seconds


- Recursive: 1.7807822227478027 seconds
- Iterative: 0.0 seconds
- Memoization: 0.0009989738464355469 seconds

Levando em consideração os resultados acima, fica evidente que a memoização otimiza o tempo de execução de funções recursivas.

## Função Recursiva: Busca Binária

In [16]:
def busca_binaria(l, x, inicio, fim):
    meio = (inicio + fim) // 2

    if fim < inicio: 
        return -1 # Retorna "-1" quando não encontra o elemento.

    if x == l[meio]:
        return meio
    
    if x < l[meio]:
        return busca_binaria(l, x, inicio, meio - 1)
    
    if x > l[meio]:
        return busca_binaria(l, x, meio + 1, fim)

Gerar lista com números aleatórios.

In [11]:
import random

Números de 0 a 99 e uma lista com 20 elementos.

In [12]:
l = random.sample(range(100), 20)
l

[85, 92, 40, 32, 3, 61, 73, 20, 75, 21, 38, 88, 52, 22, 55, 90, 71, 1, 6, 0]

Ordenar os valores com a função ".sort( )"

In [13]:
l.sort()
l

[0, 1, 3, 6, 20, 21, 22, 32, 38, 40, 52, 55, 61, 71, 73, 75, 85, 88, 90, 92]

Fazendo uma busca usando a função "busca_binaria".

In [17]:
busca_binaria(l, 40, 0, 19)

9